src/ring_theory/trace.lean

```diff
@@ -0,0 +1,129 @@
+
+We only define the trace for left multiplication (`matrix.lmul`, i.e. `algebra.lmul_left`).
+For now, the definitions assume `S` is commutative, so the choice doesn't matter anyway.
+
+## References
+
+ * https://en.wikipedia.org/wiki/Field_trace
+
+-/
+
+universes u v w
+
+variables {R S T : Type*} [comm_ring R] [comm_ring S] [comm_ring T]
+variables {A: Type*} [integral_domain A] [algebra A S]
+variables [algebra R S] [algebra R T]
+variables {K L : Type*} [field K] [field L] [algebra K L]
+variables {ι : Type w} [fintype ι]
+
+open finite_dimensional
+open linear_map
+open matrix
+
+open_locale big_operators
+open_locale matrix
+
+namespace algebra
+
+variables {b : ι → S} (hb : is_basis R b)
+
+variables (R S)
+
+/-- The trace of an element `s` of an `R`-algebra is the trace of `(*) s`,
+as an `R`-linear map. -/
+noncomputable def trace : S →ₗ[R] R :=
+(linear_map.trace R S).comp (lmul R S).to_linear_map
+
+variables {S}
+
+@[simp] lemma trace_apply (s : S) : trace R S s = linear_map.trace R S (lmul R S s) := rfl
+
+lemma trace_eq_zero_of_not_exists_basis
+  (h : ¬ ∃ s : finset S, is_basis R (λ x, x : (↑s : set S) → S)) : trace R S = 0 :=
+by { ext s, simp [linear_map.trace, h] }
+
+include hb
+
+variables {R}
+
+-- Can't be a `simp` lemma because it depends on a choice of basis
+lemma trace_eq_matrix_trace [decidable_eq ι] (hb : is_basis R b) (s : S) :
+  trace R S s = matrix.trace _ R _ (matrix.lmul hb s) :=
+by rw [trace_apply, linear_map.trace_eq_matrix_trace _ hb, to_matrix_lmul_eq]
+
+/-- If `x` is in the base field `K`, then the trace is `[L : K] * x`. -/
+lemma trace_algebra_map_of_basis (x : R) :
+  trace R S (algebra_map R S x) = fintype.card ι • x :=
+begin
+  haveI := classical.dec_eq ι,
+  rw [trace_apply, linear_map.trace_eq_matrix_trace R hb, trace_diag],
+  convert finset.sum_const _,
+  ext i,
+  simp,
+end
+omit hb
+
+/-- If `x` is in the base field `K`, then the trace is `[L : K] * x`.
+
+(If `L` is not finite-dimensional over `K`, then `trace` and `findim` return `0`.)
+-/
+@[simp]
+lemma trace_algebra_map (x : K) : trace K L (algebra_map K L x) = findim K L • x :=
+begin
+  by_cases H : ∃ s : finset L, is_basis K (λ x, x : (↑s : set L) → L),
+  { rw [trace_algebra_map_of_basis H.some_spec, findim_eq_card_basis H.some_spec] },
+  { simp [trace_eq_zero_of_not_exists_basis K H, findim_eq_zero_of_not_exists_basis H] },
+end
+
+section trace_form
+
+variables (R S)
+
+/-- The `trace_form` maps `x y : S` to the trace of `x * y`.
+It is a symmetric bilinear form and is nondegenerate if the extension is separable. -/
+noncomputable def trace_form : bilin_form R S :=
+{ bilin := λ x y, trace R S (x * y),
+  bilin_add_left := λ x y z, by rw [add_mul, linear_map.map_add],
+  bilin_smul_left := λ x y z, by rw [smul_mul_assoc, linear_map.map_smul, smul_eq_mul],
+  bilin_add_right := λ x y z, by rw [mul_add, linear_map.map_add],
+  bilin_smul_right := λ x y z, by rw [mul_smul_comm, linear_map.map_smul, smul_eq_mul], }
```

Comment:
Does it make sense to allow something like:

```suggestion
noncomputable def trace_form : bilin_form R S :=
(trace R S).comp_bilin_form (algebra.lmul R S0_.to_linear_map.to_bilin
```
I don't think `comp_bilin_form` exists right now, but its proof would be almost exactly what you have here.

Reply:
`algebra.lmul` is not exactly a bilinear form, since it is a map `S -> S -> S` instead of `S -> S -> R`. But it would make sense to define a two-argument `comp_bilin : (C -> D) -> (A -> B -> C) -> (A -> B -> D)`, which can be turned into a bilinear *form* (rather than bilinear *map*) with the usual `linear_map.to_bilin`.

Reply:
I managed to define the following but it needs a lot of "new" structure. Anyone have some applied monadology to re-use existing definitions?

```lean
variables {R M M₂ M₃ M₄ : Type*}
variables [comm_semiring R]
variables [add_comm_monoid M] [add_comm_monoid M₂] [add_comm_monoid M₃] [add_comm_monoid M₄]
variables [semimodule R M] [semimodule R M₂] [semimodule R M₃] [semimodule R M₄]

/-- Applying a linear map to the result of a bilinear map gives a bilinear map;
this construction is itself a bilinear map. -/
@[simps]
def comp_bilin : (M₃ →ₗ[R] M₄) →ₗ[R] (M →ₗ[R] M₂ →ₗ[R] M₃) →ₗ[R] (M →ₗ[R] M₂ →ₗ[R] M₄) :=
{ to_fun := λ f,
    { to_fun := λ g,
        { to_fun := λ x, f.comp (g x),
          map_add' := λ x y, by rw [g.map_add, comp_add],
          map_smul' := λ c x, by rw [g.map_smul, comp_smul] },
      map_add' := λ g g', by { ext, simp only [map_add, comp_apply, coe_mk, add_apply] },
      map_smul' := λ c g, by { ext, simp only [map_smul, comp_apply, coe_mk, smul_apply] } },
  map_add' := λ f f', by { ext, simp only [map_add, comp_apply, coe_mk, add_apply] },
  map_smul' := λ c f, by { ext, simp only [map_smul, comp_apply, coe_mk, smul_apply] } }
```

Reply:
Aren't you describing `linear_map.compr₂`, possibly with its arguments flipped?

Reply:
Good point, I didn't think to look in `tensor_algebra.lean`.